core/src/locktower.rs

```diff
@@ -0,0 +1,580 @@
+        vote_accounts: F,
+        flat_parents: &HashMap<u64, HashSet<u64>>,
+    ) -> HashMap<u64, StakeLockout>
+    where
+        F: Iterator<Item = (Pubkey, Account)>,
+    {
+        let mut stake_lockouts = HashMap::new();
+        for (key, account) in vote_accounts {
+            let lamports: u64 = *self.epoch.stakes.get(&key).unwrap_or(&0);
+            if lamports == 0 {
+                continue;
+            }
+            let mut vote_state: VoteState = VoteState::deserialize(&account.data)
+                .expect("bank should always have valid VoteState data");
+            let start_root = vote_state.root_slot;
+            vote_state.process_vote(Vote { slot: bank_slot });
+            for vote in &vote_state.votes {
+                Self::insert_fork_tree_lockouts(&mut stake_lockouts, &vote, flat_parents);
+            }
+            if start_root != vote_state.root_slot {
+                if let Some(root) = start_root {
+                    let vote = Lockout {
+                        confirmation_count: MAX_LOCKOUT_HISTORY as u32,
+                        slot: root,
+                    };
+                    Self::insert_fork_tree_lockouts(&mut stake_lockouts, &vote, flat_parents);
+                }
+            }
+            if let Some(root) = vote_state.root_slot {
+                let vote = Lockout {
+                    confirmation_count: MAX_LOCKOUT_HISTORY as u32,
+                    slot: root,
+                };
+                Self::insert_fork_tree_lockouts(&mut stake_lockouts, &vote, flat_parents);
+            }
+            // each account hash a stake for all the forks in the active tree for this bank
+            Self::insert_fork_tree_stake(&mut stake_lockouts, bank_slot, lamports, flat_parents);
+        }
+        stake_lockouts
+    }
+
+    pub fn update_epoch(&mut self, bank: &Bank) {
+        let bank_epoch = bank.get_epoch_and_slot_index(bank.slot()).0;
+        if bank_epoch != self.epoch.slot {
+            assert!(bank_epoch > self.epoch.slot, "epoch cannot move backwards");
+            self.epoch = Epoch::new_from_bank(bank);
+        }
+    }
+
+    pub fn record_vote(&mut self, slot: u64) {
+        self.lockouts.process_vote(Vote { slot });
+    }
+
+    pub fn calculate_weight(&self, stake_lockouts: &HashMap<u64, StakeLockout>) -> u128 {
+        let mut sum = 0u128;
+        let root_slot = self.lockouts.root_slot.unwrap_or(0);
+        for (slot, stake_lockout) in stake_lockouts {
+            if self.lockouts.root_slot.is_some() && *slot <= root_slot {
+                continue;
+            }
+            sum += u128::from(stake_lockout.lockout) * u128::from(stake_lockout.stake)
+        }
+        sum
+    }
+
+    pub fn check_already_voted(&self, slot: u64) -> bool {
+        for vote in &self.lockouts.votes {
+            if vote.slot == slot {
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn check_vote_lockout(
```

Comment:
This name seems odd, but no doc comment to help offer a better suggestion. Maybe . `is_locked_out`?